blocks/serialization.py

```diff
@@ -1,65 +1,181 @@
+        ``True``.
+    allow_duplicates : bool, optional
+        Allow multiple shared variables to have the same name, in which
+        case they will be numbered e.g. `x`, `x_2`, `x_3`, etc. Defaults to
+        ``True``.
+
+    Raises
+    ------
+    ValueError
+        If an unnamed shared variable is encountered and `allow_unnamed` is
+        ``False``, or if two shared variables have the same name, and
+        `allow_duplicates` is ``False``.
 
-Python is unable to serialize nested functions. Make sure that all your \
-functions are available in the global namespace."""
+    """
+    def __init__(self, zip_file, allow_unnamed=True, allow_duplicates=True):
+        super(PersistentParameterID, self).__init__(zip_file)
+        self.name_counter = defaultdict(int)
+        self.ndarray_names = {}
+        self.allow_unnamed = allow_unnamed
+        self.allow_duplicates = allow_duplicates
+
+    def _resolve_name(self, obj):
+        if id(obj) in self.ndarray_names:
+            name = self.ndarray_names[id(obj)]
+            count = self.name_counter[name]
+            if count:
+                if not self.allow_duplicates:
+                    raise ValueError("multiple shared variables with the name "
+                                     "`{0}` found".format(name))
+                name = '{0}_{1}'.format(name, count + 1)
+            self.name_counter[name] += 1
+            return name
+        return super(PersistentParameterID, self)._resolve_name(obj)
+
+    def __call__(self, obj):
+        if isinstance(obj, SharedVariable):
+            if obj.name:
+                if obj.name == 'pkl':
+                    raise ValueError(
+                        "can't pickle shared variable with name `pkl`")
+                if hasattr(obj.tag, 'annotations'):
+                    name = BRICK_DELIMITER.join(
+                        [brick.name for brick
+                         in obj.tag.annotations[0].get_unique_path()] +
+                        [obj.name])
+                else:
+                    name = obj.name
+                self.ndarray_names[id(obj.container.storage[0])] = name
+            elif not self.allow_unnamed:
+                raise ValueError("unnamed shared variable, {}".format(obj))
+        return super(PersistentParameterID, self).__call__(obj)
+
+
+class PicklerWithWarning(pickle.Pickler):
+    if six.PY2:
+        dispatch = pickle.Pickler.dispatch.copy()
+    else:
+        dispatch_table = copyreg.dispatch_table.copy()
+
+    def save_global(self, obj, **kwargs):
+        module = getattr(obj, "__module__", None)
+        if module == '__main__':
+            warnings.warn(
+                MAIN_MODULE_WARNING.format(kwargs.get('name', obj.__name__))
+            )
+        pickle.Pickler.save_global(self, obj, **kwargs)
+    if six.PY3:
+        dispatch_table[six.types.FunctionType] = save_global
+    else:
+        dispatch[six.types.ClassType] = save_global
```

Comment:
Here comes the profile

```python
         145163322 function calls (144834560 primitive calls) in 208.903 seconds

   Ordered by: cumulative time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
    953/1    0.032    0.000  214.053  214.053 {built-in method exec}
        1    0.086    0.086  214.053  214.053 dump.py:1(<module>)
        1    0.000    0.000  212.078  212.078 serialization.py:121(dump)
 312909/1    3.413    0.000  209.698  209.698 pkl_utils.py:326(zipadd)
        1    0.000    0.000  208.948  208.948 serialization.py:170(func)
        1    1.301    1.301  208.948  208.948 {method 'dump' of '_pickle.Pickler' objects}
   882956    1.569    0.000  207.632    0.000 serialization.py:80(__call__)
   882956    3.142    0.000  205.632    0.000 pkl_utils.py:144(__call__)
   312908    0.780    0.000  102.769    0.000 pkl_utils.py:147(write_array)
   312908    5.249    0.000  101.989    0.000 format.py:507(write_array)
   312908    6.873    0.000   93.054    0.000 format.py:270(_write_array_header)
   312908    7.256    0.000   74.038    0.000 format.py:415(_filter_header)
   312909   11.094    0.000   53.897    0.000 zipfile.py:1315(write)
  5632344   20.615    0.000   49.981    0.000 tokenize.py:473(_tokenize)
   312909    1.500    0.000   27.418    0.000 tempfile.py:571(NamedTemporaryFile)
   312910    2.044    0.000   21.790    0.000 tempfile.py:330(_mkstemp_inner)
   312908    0.619    0.000   14.401    0.000 tokenize.py:302(untokenize)
  5008285   13.965    0.000   13.965    0.000 {method 'match' of '_sre.SRE_Pattern' objects}
   312908    8.048    0.000   13.382    0.000 tokenize.py:245(untokenize)
...
```

The actual reason why it is so slow is unexpected: the log contains lots of small numpy arrays! So I think you guys should only apply the persistent id magic to the shared variables.